diff --git a/Android.bp b/Android.bp
index e15ddc5..4e63b88 100644
--- a/Android.bp
+++ b/Android.bp
@@ -53,11 +53,7 @@ rust_library {
     cargo_pkg_version: "0.17.1",
     srcs: ["src/lib.rs"],
     edition: "2018",
-    features: [
-        "boringssl-vendored",
-        "default",
-    ],
-    rustlibs: [
+    rlibs: [
         "liblazy_static",
         "liblibc",
         "liblibm",
@@ -67,7 +63,8 @@ rust_library {
         "libslab",
         "libsmallvec",
     ],
-    static_libs: [
+    prefer_rlib: true,
+    shared_libs: [
         "libcrypto",
         "libssl",
     ],
@@ -80,7 +77,7 @@ rust_library {
     min_sdk_version: "29",
 }
 
-rust_ffi_shared {
+rust_ffi {
     name: "libquiche_ffi",
     host_supported: true,
     crate_name: "quiche",
@@ -88,11 +85,7 @@ rust_ffi_shared {
     cargo_pkg_version: "0.17.1",
     srcs: ["src/lib.rs"],
     edition: "2018",
-    features: [
-        "boringssl-vendored",
-        "default",
-    ],
-    rustlibs: [
+    rlibs: [
         "liblazy_static",
         "liblibc",
         "liblibm",
@@ -102,7 +95,8 @@ rust_ffi_shared {
         "libslab",
         "libsmallvec",
     ],
-    static_libs: [
+    prefer_rlib: true,
+    shared_libs: [
         "libcrypto",
         "libssl",
     ],
@@ -143,7 +137,7 @@ rust_test_host {
         "libsmallvec",
         "liburl",
     ],
-    static_libs: [
+    shared_libs: [
         "libcrypto",
         "libssl",
     ],
@@ -190,6 +184,7 @@ rust_test {
         "examples/cert-big.crt",
         "examples/rootca.crt",
     ],
+    shared_libs: ["libc++"],
 }
 
 rust_library_rlib {
@@ -200,10 +195,6 @@ rust_library_rlib {
     cargo_pkg_version: "0.17.1",
     srcs: ["src/lib.rs"],
     edition: "2018",
-    features: [
-        "boringssl-vendored",
-        "default",
-    ],
     rustlibs: [
         "liblazy_static",
         "liblibc",
